File: YJ/20200904/BinSearch/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool is_possible(long long mid, int n, vector<int> times) {
    long long sum = 0;
    for (int i = 0; i < times.size(); i ++) {
        sum += mid / times[i];
    }
    if(sum >= n) return true;
    else return false;
}

long long solution(int n, vector<int> times) {
    long long answer = 0, lft, rt, mid;
    
    sort(times.begin(), times.end());
    
    rt = (long long)times[times.size() - 1] * n;
    lft = 1;
    
    while(lft <= rt) {
        mid = (lft + rt) / 2;
        
        if (is_possible(mid, n, times)) {
            rt = mid - 1;
            answer = mid;
        } else {
            lft = mid + 1;
        }
    }
    
    return answer;
}

File: JY/8/8.28/입국심사.py
def solution(n, times):
    ans = 0
    m = min(times)
    start = 0
    end = m*n
    
    while start<end:
        mid = (start+end)//2
        cnt = 0
        for t in times:
            cnt += mid//t
        if cnt>=n:
            end = mid
        else:
            start= mid+1
        
    ans = end
    return ans